src/Data/Nat/Bounded.agda

```diff
@@ -0,0 +1,69 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Bounded natural numbers, and their equivalence to `Fin`
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Data.Nat.Bounded where
+
+open import Data.Nat.Base as ℕ
+import Data.Nat.Properties as ℕ
+open import Data.Fin.Base as Fin using (Fin; zero; suc; toℕ)
+import Data.Fin.Properties as Fin
+open import Function.Base using (id; _∘_; _on_)
+open import Relation.Binary.Core using (_⇒_)
+open import Relation.Binary.PropositionalEquality.Core as ≡ using (_≡_; refl)
+open import Relation.Nullary.Decidable.Core using (recompute)
+open import Relation.Nullary.Negation.Core using (¬_)
+
+private
+  variable
+    m n : ℕ
+
+------------------------------------------------------------------------
+-- Definition
+
+record BoundedNat (n : ℕ) : Set where
+  constructor ⟦_⟧<
+  field
+    value : ℕ
+    .{{bounded}} : value < n
```

Comment:
Well, that had been my thinking underlying #1998 and (the reverted parts of) #2000 ... but indeed reimplementing this via `Data.Refinement` wouldn't introduce that extra twist. I had the ... even 'gut feeling' seems a bit strong ... sense that in suitable deployment contexts (fixed-width arithmetic; positional notation for numerals; ...) it might be the case the ambient instances would be available from the context ... but perhaps you're correct...

Reply:
I like the idea of adding `BoundedNat` as an equivalent for `Fin`, but I think the 'experiment' of using an instance for `bounded` is premature.

Supporting evidence: in code that Conor and I were playing around with, we had exactly `BoundedNat` in our extensional universe as our definition of `Fin`, and we did have to supply the proof a non-trivial number of times. (That the proof was irrelevant without needed Agda's irrelevance machinery was what we were really after.)

Reply:
OK, so that's two of you... so either:
* I remove the `instance` bracketing (easy);
* I go all-out for `T (value <ᵇ n)` as above (more involved, but most of the machinery is already in #2000 )

I guess it's the first choice, then, with the second as a possible downstream refactoring?

Reply:
NB making `bounded` an 'ordinary' (even if `irrelevant`) field vitiates somewhat my choice of `⟦_⟧<` as a constructor, which could be used 'bare', without bracketing, as a pattern constructor, in favour of `⟦_⟧<_` with the explicit bound argument, and the `value` argument position now mostly redundant... oh well. The redundancy is harmless, unless someone else has a compelling improvement on my notation?

Reply:
The addition of a `Number` instance with `Constraint m = T (m <ᵇ n)` biases towards the second alternative...
... but `stdlib` style suggests using `True (m <? n)` instead... cf. `Data.Fin.Literals`

Any comment on the trade-offs between the two, given that the latter is much more indirect than the former...?